riscv/mmu.cc

```diff
@@ -531,18 +531,26 @@ reg_t mmu_t::walk(mem_access_info_t access_info)
       base = ppn << PGSHIFT;
     } else if ((pte & PTE_U) ? s_mode && (type == FETCH || !sum) : !s_mode) {
       break;
-    } else if (!(pte & PTE_V) || (!(pte & PTE_R) && (pte & PTE_W))) {
+    } else if (!(pte & PTE_V) || (!(pte & PTE_R) && (pte & PTE_W) && !access_info.flags.shstk)) {
```

Comment:
Various aspects are missing:
All of below should be based on effective privilege of the memory access and not current privilege.
2. If virtual memory system is Bare then ss load and store cause SAMO access-fault
3. SS load and store cause SAMO access fault on access to RW, RWX, X, RX pages
4. SS load and store cause SAMO page fault on access to RO, invalid, and malformed PTEs
5. Non-SS store to SS pages cause SAMO access fault
6. Instruction fetches to SS pages cause instruction access fault

SS encodings in PTEs stay reserved if SS extension is not active.

Reply:
Thank you for explaining the rule in such detail.
The rules for checking the PTE rule are a bit confusing, and I want to separate each action independently.

Only check whether the permission of the SS page is VW like following code.
```
else if (check_ss && ((pte & vw) != vw)) {
      // ss load/store page with R=0, W=1, X=0 and valid
      break;
```
After this, check the PTE permission of the SS page for invalidity.
```
  if (check_ss) {
    if (!(invalid_pte & PTE_V))
      throw trap_store_page_fault(virt, addr, 0, 0);
    else if (invalid_pte & PTE_W || invalid_pte & PTE_X)
      throw trap_store_access_fault(virt, addr, 0, 0);
    else
      throw trap_store_page_fault(virt, addr, 0, 0);
  }
```

Reply:
I suggest the pattern used by CMO and AMOs to convert load traps to store traps - please see `convert_load_traps_to_store_traps`. in the ss_load we can use this wrapper to convert the load*fault to store*fault.

Reply:
I think we could do this:
```
+  bool ss_access = access_info.ss_access;
+
+  if (ss_access) {
+    if (vm.levels == 0)
+        throw trap_store_access_fault(virt, addr, 0, 0);
+    type = STORE;
+  }
+
   if (vm.levels == 0)
     return s2xlate(addr, addr & ((reg_t(2) << (proc->xlen-1))-1), type, type, virt, hlvx, false) & ~page_mask; // zero-extend from xlen
 
@@ -516,6 +524,7 @@ reg_t mmu_t::walk(mem_access_info_t access_info)
     reg_t ppn = (pte & ~reg_t(PTE_ATTR)) >> PTE_PPN_SHIFT;
     bool pbmte = virt ? (proc->get_state()->henvcfg->read() & HENVCFG_PBMTE) : (proc->get_state()->menvcfg->read() & MENVCFG_PBMTE);
     bool hade = virt ? (proc->get_state()->henvcfg->read() & HENVCFG_ADUE) : (proc->get_state()->menvcfg->read() & MENVCFG_ADUE);
+    bool sse = virt ? (proc->get_state()->henvcfg->read() & HENVCFG_SSE) : (proc->get_state()->menvcfg->read() & MENVCFG_SSE);
 
     if (pte & PTE_RSVD) {
       break;
@@ -531,11 +540,21 @@ reg_t mmu_t::walk(mem_access_info_t access_info)
       base = ppn << PGSHIFT;
     } else if ((pte & PTE_U) ? s_mode && (type == FETCH || !sum) : !s_mode) {
       break;
-    } else if (!(pte & PTE_V) || (!(pte & PTE_R) && (pte & PTE_W))) {
+    } else if (!(pte & PTE_V) ||
+              (!(pte & PTE_R) && (pte & PTE_W) && ((!sse && !(pte & PTE_X)) || (pte & PTE_X)))) {
       break;
+    } else if ((!(pte & PTE_R) && (pte & PTE_W) && !(pte & PTE_X)) && (type == STORE && !ss_access)) {
+      // non-shadow-stack stores to shadow stack pages cause store access-fault
+      throw trap_store_access_fault(virt, addr, 0, 0);
+    } else if ((!(pte & PTE_R) && (pte & PTE_W) && !(pte & PTE_X)) && type == FETCH) {
+      // fetch from shadow stack pages cause instruction access-fault
+      throw trap_instruction_access_fault(virt, addr, 0, 0);
+    } else if ((((pte & PTE_R) && (pte & PTE_W)) || (pte & PTE_X)) && ss_access) {
+      // shadow stack access cause store access fault if xwr!=010 and xwr!=001
+      throw trap_store_access_fault(virt, addr, 0, 0);
     } else if (type == FETCH || hlvx ? !(pte & PTE_X) :
                type == LOAD          ? !(pte & PTE_R) && !(mxr && (pte & PTE_X)) :
-                                       !((pte & PTE_R) && (pte & PTE_W))) {
+ 
```

Reply:
This should also be added to pmp_ok:
```
if (ss_access)
   type = STORE;
```

Reply:
Thanks for your narrow down rule.
I add some more detailed descriptions and remove the STORE check in the end
```
else if (type == FETCH || hlvx ? !(pte & PTE_X) :
               type == LOAD && !(pte & PTE_R) && !(mxr && (pte & PTE_X))) {
```